PDF stamping of approved invoices

theskumar opened the pull request:
- [x] Add approved by staff to the approved invoice pdfs
- [x] Add inovice approver and date

![Screenshot 2024-05-20 at 6  33 05@2x](https://github.com/HyphaApp/hypha/assets/236356/9c9150a7-33d1-41fe-98e1-75e4ef0646f2)

Closes: https://github.com/HyphaApp/hypha/issues/3539

theskumar commented:
@sandeepsajan0 @frjo Are you good with denormalizing and add the `approved_at` and `approved_by` to the invoice model? 

The code to update them will go in here: 

https://github.com/HyphaApp/hypha/blob/1b401334914f013b7d30e9bf4c2838361ee7c8f7/hypha/apply/projects/views/payment.py#L81-L87

theskumar commented:
Added the logic to store the approver and date at https://github.com/HyphaApp/hypha/pull/3698/commits/f426d46f5be088b3228dca94c155789ceb28128c

Let me know if this something that need to be improved upon? e.g. displaying both finance1 and finance2 info etc.

sandeepsajan0 commented on hypha/apply/projects/models/payment.py:
```diff
@@ -183,6 +183,13 @@ class Invoice(models.Model):
+        null=True,
+        blank=True,
+    )
+    approved_at = models.DateTimeField(null=True, blank=True)
```
Can we rename it to `final_approval_...` or something that doesn't confuse it with internal approvals like staff approval or finance approval(in case of extended workflow)? Because we have to stamp the pdf only with final approval(finance approval or finance2 approval).

Or we may directly look out for the final approval activity as the [activity](https://github.com/HyphaApp/hypha/blob/main/hypha/apply/activity/models.py#L98) already has the date and user.

sandeepsajan0 commented on hypha/apply/projects/views/payment.py:
```diff
@@ -84,6 +86,9 @@ def form_valid(self, form):
             and self.request.user.is_finance_level_1
             and self.object.status == APPROVED_BY_FINANCE
         ):
+            self.object.approved_by = self.request.user
```
It is not the right place to update approved_by or approved_at because here we are just notifying the finance members to approve the invoice.

It requires adding another condition to check if the status is `Approved by Finance`(without extended).

theskumar commented:
@sandeepsajan0 thanks for feedback. I would like wait for the feedback on the content sample itself and then tackle both at once. 

- https://github.com/HyphaApp/hypha/issues/3539#issuecomment-1860480909

theskumar committed:
Add invoice approval stamps

Attach additional page to invoice if the invoices are approved by staff,
finance or other staff members

theskumar commented:
Updated the code to make use of activity feed. It will list all the approvals with timestamps.

sandeepsajan0 commented on hypha/apply/projects/views/payment.py:
```diff
@@ -416,12 +419,35 @@ def dispatch(self, *args, **kwargs):
+                related_content_type__model="invoice",
+                related_object_id=self.invoice.id,
+                message__icontains="Approved by",
+            ).visible_to(self.request.user):
```
It will pick all the Approved by ... statuses. If an invoice has been back and forth between finance and staff then it might show multiple `Approved by Staff` but I think only latest one should be fine. We might need distinct activities.  Also, we don't want to add this for invoices with status `Approved by finance`?  And we might need to hide it from the applicant.
![image](https://github.com/HyphaApp/hypha/assets/23638629/d091a993-91c3-4169-871d-eccf5573b036)

sandeepsajan0 commented on hypha/apply/projects/views/payment.py:
```diff
@@ -121,6 +123,7 @@ def form_valid(self, form):
             and self.request.user.is_finance_level_1
             and self.object.status == APPROVED_BY_FINANCE
         ):
+            self.object.save()
```
I think it was getting saved in `form_valid()` in line100?
